src/model_theory/basic.lean

```diff
@@ -0,0 +1,317 @@
+  `n`-ary relations.
+* A `first_order.language.Structure` interprets the symbols of a given `first_order.language` in the
+  context of a given type.
+* A `first_order.language.hom`, denoted `M →[L] N`, is a map from the `L`-structure `M` to the
+  `L`-structure `N` that commutes with the interpretations of functions, and which preserves the
+  interpretations of relations (although only in the forward direction).
+* A `first_order.language.embedding`, denoted `M ↪[L] N`, is an embedding from the `L`-structure `M`
+  to the `L`-structure `N` that commutes with the interpretations of functions, and which preserves
+  the interpretations of relations in both directions.
+* A `first_order.language.equiv`, denoted `M ≃[L] N`, is an equivalence from the `L`-structure `M`
+  to the `L`-structure `N` that commutes with the interpretations of functions, and which preserves
+  the interpretations of relations in both directions.
+
+## References
+For the Flypitch project:
+- [J. Han, F. van Doorn, *A formal proof of the independence of the continuum hypothesis*]
+[flypitch_cpp]
+- [J. Han, F. van Doorn, *A formalization of forcing and the unprovability of
+the continuum hypothesis*][flypitch_itp]
+
+-/
+
+namespace first_order
+
+/-- A first-order language consists of a type of functions of every natural-number arity and a
+  type of relations of every natural-number arity. -/
+structure language :=
+(functions : ℕ → Type*) (relations : ℕ → Type*)
+
+namespace language
+
+/-- The empty language has no symbols. -/
+def empty : language := ⟨λ _, pempty, λ _, pempty⟩
+
+instance : inhabited language := ⟨empty⟩
+
+/-- The type of constants in a given language. -/
+@[nolint has_inhabited_instance] def const (L : language) := L.functions 0
+
+variable (L : language)
+
+/-- A language is relational when it has no function symbols. -/
+class is_relational : Prop :=
+(empty_functions : ∀ n, L.functions n → false)
+
+/-- A language is algebraic when it has no relation symbols. -/
+class is_algebraic : Prop :=
+(empty_relations : ∀ n, L.relations n → false)
+
+variable {L}
+
+instance is_relational_of_empty_functions {symb : ℕ → Type*} : is_relational ⟨λ _, pempty, symb⟩ :=
+⟨by { intro n, apply pempty.elim }⟩
+
+instance is_algebraic_of_empty_relations {symb : ℕ → Type*}  : is_algebraic ⟨symb, λ _, pempty⟩ :=
+⟨by { intro n, apply pempty.elim }⟩
+
+instance is_relational_empty : is_relational (empty) := language.is_relational_of_empty_functions
+instance is_algebraic_empty : is_algebraic (empty) := language.is_algebraic_of_empty_relations
+
+variables (L) (M : Type*)
+
+/-- A first-order structure on a type `M` consists of interpretations of all the symbols in a given
+  language. Each function of arity `n` is interpreted as a function sending tuples of length `n`
+  (modeled as `(fin n → M)`) to `M`, and a relation of arity `n` is a function from tuples of length
+  `n` to `Prop`. -/
+class Structure :=
```

Comment:
I'm taking this convention from Flypitch, but it's because `structure` is very much taken. If I am careful with namespaces, I can get `first_order.language.structure` to work, but then after that, if I'm in the `first_order.language` namespace, I can't use the regular Lean `structure` command.

If you prefer, we could call this a `struct` or something.

Reply:
Oh, yes. :-) This is fine.